mea/src/mutex/mod.rs

```diff
@@ -314,6 +324,125 @@ impl<T: ?Sized> DerefMut for MutexGuard<'_, T> {
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// use mea::mutex::{Mutex, MutexGuard};
+    ///
+    /// #[derive(Debug)]
+    /// struct User {
+    ///     id: u32,
+    ///     profile: UserProfile,
+    /// }
+    ///
+    /// #[derive(Debug)]
+    /// struct UserProfile {
+    ///     email: String,
+    ///     name: String,
+    /// }
+    ///
+    /// let user = User {
+    ///     id: 1,
+    ///     profile: UserProfile {
+    ///         email: "user@example.com".to_owned(),
+    ///         name: "Alice".to_owned(),
+    ///     },
+    /// };
+    ///
+    /// let mutex = Mutex::new(user);
+    /// let guard = mutex.lock().await;
+    ///
+    /// // Map to only access the user's profile, allowing fine-grained locking
+    /// let profile_guard = MutexGuard::map(guard, |user| &mut user.profile);
+    /// assert_eq!(profile_guard.email, "user@example.com");
+    /// # }
+    /// ```
+    pub fn map<U, F>(mut orig: Self, f: F) -> MappedMutexGuard<'a, U>
+    where
+        F: FnOnce(&mut T) -> &mut U,
+        U: ?Sized,
+    {
+        let d = NonNull::from(f(&mut *orig));
+        let orig = ManuallyDrop::new(orig);
+        MappedMutexGuard {
+            d,
+            s: &orig.lock.s,
+            variance: PhantomData,
+        }
+    }
+
+    /// Attempts to make a new [`MappedMutexGuard`] for a component of the locked data. The
+    /// original guard is returned if the closure returns `None`.
+    ///
+    /// This operation cannot fail as the `MutexGuard` passed in already locked the mutex.
+    ///
+    /// This is an associated function that needs to be used as `MutexGuard::try_map(...)`. A
+    /// method would interfere with methods of the same name on the contents of the locked data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// use mea::mutex::{Mutex, MutexGuard};
+    ///
+    /// #[derive(Debug)]
+    /// struct Database {
+    ///     users: std::collections::HashMap<u32, String>,
+    ///     admin_user_id: Option<u32>,
+    /// }
+    ///
+    /// let mut db = Database {
+    ///     users: std::collections::HashMap::new(),
+    ///     admin_user_id: Some(1),
+    /// };
+    /// db.users.insert(1, "admin@example.com".to_owned());
+    ///
+    /// let mutex = Mutex::new(db);
+    /// let guard = mutex.lock().await;
+    ///
+    /// // Try to map to admin user's email if admin exists
+    /// let admin_email_guard = MutexGuard::try_map(guard, |db| {
+    ///     if let Some(admin_id) = db.admin_user_id {
+    ///         db.users.get_mut(&admin_id)
+    ///     } else {
+    ///         None
+    ///     }
+    /// }).expect("admin user should exist");
+    ///
+    /// assert_eq!(&*admin_email_guard, "admin@example.com");
+    /// # }
+    /// ```
+    pub fn try_map<U, F>(mut orig: Self, f: F) -> Result<MappedMutexGuard<'a, U>, Self>
```

Comment:
Although I share the same understanding with your rationale in PR description, I tend to use `filter_map` to provide users a familiar experience from sync mutexes.